Log volume process failure event to pod.

tsmetana opened the pull request:
When volume manager or attach/detach controller start to process
a volume and encounter errors, there is no error event recorded
in Pod so that user have to wait until attach time out to realize
that something wrong with the volume. Besides, the time out event
does not show the detail error information to the user.

Need to log volume process failure event to pod to show user the
detail volume handling error immediately once the error happens.

Fixes #58272

Note to reviewers:
This is only a rebased PR #58273 by @xingzhou -- the original PR looks to be stalled, but we're still interested in the functionality so I'm re-trying to push it through again.

There is one aspect I'm still quite unsure about: it adds a new kubelet option to configure the interval to configure the failure event emission interval... I don't know of a better way to do this but adding new kubelet option is really an invasive change. I would prefer to remove that part.

```release-note
kubelet: add the ProcessVolumeFailureEventInterval configuration option that can be used to control the interval between reporting volume processing failure events.
```

xingzhou committed:
Log volume process failure event to pod.

When volume manager or attach/detach controller start to process
a volume and encounter errors, there is no error event recorded
in Pod so that user have to wait until attach time out to realize
that something wrong with the volume. Besides, the time out event
does not show the detail error information to the user.

Need to log volume process failure event to pod to show user the
detail volume handling error immediately once the error happens.

neolit123 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -458,6 +475,27 @@ func (dswp *desiredStateOfWorldPopulator) markPodProcessed(
+	if !found || time.Since(lastTime) > dswp.failureEventInterval {
+		msgMap[msg] = time.Now()
+		dswp.recorder.Eventf(pod, v1.EventTypeWarning, kevents.FailedProcessVolume, msg)
+		glog.Errorf(logMsg)
```
is it a good idea to use verbosity level here:
`glog.V(xx).Errorf(logMsg)` ?

my assumption is that this can happen often in certain conditions, thus the possibility of logging a lot here?

neolit123 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -467,6 +505,28 @@ func (dswp *desiredStateOfWorldPopulator) deleteProcessedPod(
+
+	for podName := range dswp.failedProcessPodMessages.messages {
+		terminated, found := currentPodStateMap[podName]
+
```
no need for this empty line.

neolit123 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -195,13 +214,16 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 		}
 		dswp.processPodVolumes(pod, mountedVolumesForPod, processedVolumesForFSResize)
 	}
+
```
this empty line can be omitted.

neolit123 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -97,10 +101,14 @@ func NewDesiredStateOfWorldPopulator(
 		pods: processedPods{
 			processedPods: make(map[volumetypes.UniquePodName]bool)},
+		failedProcessPodMessages: failedProcessPodMessages{
+			messages: make(map[volumetypes.UniquePodName]map[string]time.Time)},
```
gofmt might complain about this.

neolit123 reviewed (commented):
@tsmetana thank you for digging out this PR.

i've added a couple of comments in the code.

please add a release note in the lines of:
```
kubelet: add the ProcessVolumeFailureEventInterval flag that can be used to control the interval between reporting volume processing failure events.
```

/kind feature
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
/cc @kubernetes/sig-node-pr-reviews

neolit123 commented:
@tsmetana 
there are some failing unit tests here:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/66897/pull-kubernetes-bazel-test/54616/

tsmetana commented:
@neolit123 thanks for the review. The failed tests are caused by the API change. I'll take a look at those but still -- I'm not sure about that part.

tsmetana commented:
@msau42 -- I know you asked for the interval to be configurable... I wonder if it really is necessary. Especially since there is no simple way to do it and API change with new kubelet option seems to be really drastic change to me. Moreover changing the defaults means restarting kubelet which looks to be really inconvenient way to tweak the setting.

msau42 commented:
This has the potential to be a very noisy event.  I suppose users can downgrade their nodes if it is causing problems.

Does the new dynamic kubelet config help simplify it in any way?

tsmetana commented:
@msau42 I was not aware of the dynamic kubelet configuration but after a brief scan of the documentation this looks really like a solution. We may not need the command line option which should make the patch more likely to be accepted. I need to learn more but it looks promising.

Thanks a lot.

tsmetana commented:
I've removed the command line option since the new dynamic configuration can be used to set the parameter value. I have addressed the @neolit123's comments (hopefully) and tried to fix the failing tests.

tsmetana commented:
/retest

tsmetana commented:
/retest

dashpole commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -305,6 +305,11 @@ type KubeletConfiguration struct {
+	// E.g. when configuring a wrong FlexVolume name in pod, "no volume plugin matched" error would be thrown out.
+	// Volume manager can generate a lot of similar errors because of its desired state of world loop.
+	// User can config the interval of error events reported to pod by this kubelet flag.
+	ProcessVolumeFailureEventInterval metav1.Duration
```
What meaningful decision are operators making here about how the kubelet should be configured?  We generally avoid adding "tuning" variables to the kubelet and strongly prefer just making the kubelet do the right thing by default.

msau42 replied:
I made the suggestion of making the throttling period configurable, because this event has the chance to be high frequency and we might not get the interval right and may want a way to turn it off in production.  But I agree this is a lot of instrumentation for this one message.  Is there an alternative way we could try to propagate this error up?

dashpole replied:
Maybe go with something super conservative to begin with?  1 event per minute?  That to me is better than a new configuration.

dashpole replied:
I am also curious why the errors could be so frequent.  If we fail to mount a volume, how often do we retry?  I realize shorter periods mean less latency, but we shouldn't encounter failures that often, right?  The kubelet's housekeeping sync loop is every 30 seconds, which seems to work fine as a retry rate, and is infrequent enough that we can omit events for errors we encounter.  Am I missing something?

msau42 replied:
This error is generated as part of the volume manager desired world populator loop, which is separate from kubelet's main sync loop.  It runs every [100ms](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/kubelet/volumemanager/volume_manager.go#L61)

dashpole reviewed (commented):
WRT kubelet config, we shouldn't add a kubelet parameter, but instead choose a sensible default and stick with that for all deployments.  I don't think there is a meaningful decision to be made there.

In general, we can't expect to output each error the kubelet encounters as an event.  There will always be some cases which require looking through logs for errors.  This feels like a case that is too detailed for events to make sense.
